### ISSUE TITLE
Domain memories

### ISSUE THREAD

**Issue by pdhoolia** [pdhoolia/se-agent#43]
Lack of domain understanding may sometime lead to suggestions that are not feasible or practical. For example, suggesting a change in an emulator (that is supposed to exactly replicate quantum hardware) may not be the right suggestion to fix a bug in qiskit. Or something that is not possible in the current version of the programming language, or suggesting a change that is not in line with the project's coding standards.

When agent makes such proposals, other developers may point out the issues with the suggestion via comments. The agent should be able to understand these comments and their context and learn from them. It should be able to create a succinct, condensed memory of this understanding and store it for future reference.

Here's a high-level plan to implement this feature:

- Before starting to process an issue comment, the agent may analyze the comment to check if it represents a domain understanding correction. This may be done by checking if the comment may be semantically categorized as a domain characteristic. Note: The agent probably uses the term `project` (for domain).

- If comment is identified as a domain understanding correction, the agent should add new functionality to generate a condensed memory of this understanding.
    - It should use the issue conversation context to generate this memory.
    - It should add a new LLM task for this purpose, form a prompt, and call the LLM to generate memories (there could be more than 1). Each memory should be condensed (e.g., no more than 20 tokens)
    - Agent should store these memories in a memory store. For the first version, we can use a json file in the metadata folder for storing memories.
    - Each memory should be stored as a json object with the following fields:
        - `memory_id`: An id (generated) for the memory
        - `issue_id`: The issue id in the context of which this memory was generated
        - `comment_id`: The comment id based on which this memory was generated
        - `memory`: The memory text
        - `user_id`: The user id who made the comment
        - `created_at`: The timestamp when this memory was created
        - `updated_at`: The timestamp when this memory was last updated
        - `status`: At the time of creation, the status of the memory should be `new`. This status may be updated later based on curation by an expert. Expert may mark it as `accepted`, `rejected`, `needs_review` etc.

- The agent should retrieve these memories when processing new issues or comments. In particular, it should use them in the prompts for localization and suggestions.

**Comment by se-agent** [comment]
To implement the feature of creating and managing domain memories, we need to make several modifications across multiple files. Here's a breakdown of the necessary changes:

1. **Add a new LLM task for memory generation:**

   In `se_agent/llm/api.py`, introduce a new task for memory generation. This will involve adding a new task name to the `TaskName` enum and implementing a method in `call_llm_for_task` to handle this task.

   ```python
   # Add a new task in TaskName enum
   class TaskName:
       ...
       MEMORY_GENERATION = "memory_generation"

   # Update call_llm_for_task to handle the new task
   def call_llm_for_task(task_name: TaskName, messages: list, **kwargs):
       ...
       elif task_name == TaskName.MEMORY_GENERATION:
           # Implement logic for memory generation task
           return llm.invoke(transform_to_langchain_base_chat_model_format(messages))
       ...
   ```

2. **Detect domain understanding corrections:**

   In `se_agent/issue_analyzer.py`, implement logic to detect comments that represent domain understanding corrections. This might involve semantic analysis or keyword matching.

   ```python
   def is_domain_correction(comment):
       # Implement logic to check if a comment is a domain understanding correction
       # For example, using keywords or LLM-based semantic analysis
       return "domain correction keyword" in comment['content']
   ```

3. **Generate and store memories:**

   In `se_agent/listener_core.py`, update the `process_issue_event` function to generate memories if a domain correction is identified. Use the `call_llm_for_task` function to generate memories and store them in a JSON file.

   ```python
   def process_issue_event(project: Project, issue_details, comment_details=None):
       ...
       if comment_details and is_domain_correction(comment_details):
           # Generate memories using LLM
           memory_messages = [...]  # Prepare messages for memory generation
           memories = call_llm_for_task(TaskName.MEMORY_GENERATION, memory_messages)

           # Store memories in a JSON file
           memory_store_path = os.path.join(project.metadata_folder, 'domain_memories.json')
           with open(memory_store_path, 'a') as f:
               for memory in memories:
                   memory_data = {
                       "memory_id": generate_memory_id(),
                       "issue_id": issue_details['number'],
                       "comment_id": comment_details['id'],
                       "memory": memory,
                       "user_id": comment_details['user']['login'],
                       "created_at": current_timestamp(),
                       "updated_at": current_timestamp(),
                       "status": "new"
                   }
                   json.dump(memory_data, f)
                   f.write('\n')
       ...
   ```

4. **Retrieve and utilize memories:**

   In `se_agent/listener_core.py`, update the logic to retrieve and utilize stored memories when processing new issues or comments.

   ```python
   def process_issue_event(project: Project, issue_details, comment_details=None):
       ...
       # Load and utilize domain memories
       memory_store_path = os.path.join(project.metadata_folder, 'domain_memories.json')
       if os.path.exists(memory_store_path):
           with open(memory_store_path, 'r') as f:
               memories = [json.loads(line) for line in f.readlines()]
               # Filter and utilize relevant memories in processing
               relevant_memories = [m['memory'] for m in memories if m['status'] == 'accepted']
               # Use these memories in the localization and change suggestion prompts
       ...
   ```

These changes introduce a mechanism to detect domain correction comments, generate condensed memories using an LLM, store them persistently, and leverage them in future issue processing to improve the quality of suggestions.
